src/OnChainSurplus.py

```diff
@@ -0,0 +1,113 @@
+gpv2Abi = json.loads(abiJson)
+#connecting to ethereum node
+w3 = Web3(Web3.HTTPProvider(f"https://mainnet.infura.io/v3/{INFURA_KEY}"))
+#creating an instance of GPv2 contract 
+contractInstance = w3.eth.contract(address=address, abi=gpv2Abi)
+
+class DecodedSettlement:
+    def __init__(self, tokens: List[str], clearing_prices: List[int], trades: List[Tuple[int, int, str, int, int, int, bytes, int, int, int, bytes]], interactions: List[List[Tuple[str, int, bytes]]]):
+        self.tokens = tokens
+        self.clearing_prices = clearing_prices
+        self.trades = trades
+        self.interactions = interactions
+
+    @classmethod
+    def new(cls, contractInstance, transaction):
+        # Decode the function input
+        decoded_input = contractInstance.decode_function_input(transaction.input)[1:]
+        # Convert the decoded input to the expected types
+        tokens = decoded_input[0]['tokens']
+        clearing_prices = decoded_input[0]['clearingPrices']
+        trades = decoded_input[0]['trades']
+        interactions = decoded_input[0]['interactions']
+
+        # Create and return a new instance of DecodedSettlement
+        return cls(tokens, clearing_prices, trades, interactions)
+
+
+def calculateSurplus(startBlock = None, endBlock = None, settlementHash = None):
+    if settlementHash is not None:
+        decoder(settlementHash)
+    elif startBlock is not None and endBlock is not None:
+    # define the filter criteria
+        filter_criteria = {
+            'fromBlock': int(startBlock),
+            'toBlock': int(endBlock),
+            'address': address,
+        }
+        transactions = w3.eth.filter(filter_criteria).get_all_entries()
+        transactionHashes = []
+        for tx in transactions:
+            txHash = (tx['transactionHash']).hex()
+            if txHash not in transactionHashes:
+                transactionHashes.append(txHash)
+        transactionHashes.reverse()
+        # At this point we have all the needed hashes
+
+        for hashCounter in range(len(transactionHashes)):
+            singleHash = transactionHashes[hashCounter]
+            decoder(singleHash)
+
+
+def decoder(settlementHash):
+    encodedTransaction = w3.eth.get_transaction(settlementHash)
+    decoded_settlement = DecodedSettlement.new(contractInstance, encodedTransaction)
+    # there can be multiple orders/trades in a single settlement
+    for ordersCount in range(len(decoded_settlement.trades)):
+        individualOrder = decoded_settlement.trades[ordersCount]
+        surplus = calculationFunction(decoded_settlement, individualOrder)
+        print(surplus)
+
+        endpointDataJson = requests.get(f"https://api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{settlementHash}")
+        endpointDataPy = json.loads(endpointDataJson.text)
```

Comment:
There is no need to call this function for every order individually, so it should be called outside the `for` loop. The same goes for `bucketResponseJson` and `auctionId`. All of those are the same for a single `settlementHash`.